x-pack/filebeat/input/awscloudwatch/cloudwatch.go

```diff
@@ -14,61 +14,69 @@ import (
 	awssdk "github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs"
 
-	awscommon "github.com/elastic/beats/v7/x-pack/libbeat/common/aws"
 	"github.com/elastic/elastic-agent-libs/logp"
 )
 
 type cloudwatchPoller struct {
-	numberOfWorkers      int
-	apiSleep             time.Duration
+	config               config
 	region               string
-	logStreams           []*string
-	logStreamPrefix      string
-	startTime            int64
-	endTime              int64
-	workerSem            *awscommon.Sem
 	log                  *logp.Logger
 	metrics              *inputMetrics
 	workersListingMap    *sync.Map
 	workersProcessingMap *sync.Map
+
+	// When a worker is ready for its next task, it should
+	// send to workRequestChan and then read from workResponseChan.
+	// The worker can cancel the request based on other context
+	// cancellations, but if the write succeeds it _must_ read from
+	// workResponseChan to avoid deadlocking the main loop.
+	workRequestChan  chan struct{}
+	workResponseChan chan workResponse
+
+	workerWg sync.WaitGroup
+}
+
+type workResponse struct {
+	logGroup           string
+	startTime, endTime time.Time
 }
 
 func newCloudwatchPoller(log *logp.Logger, metrics *inputMetrics,
-	awsRegion string, apiSleep time.Duration,
-	numberOfWorkers int, logStreams []*string, logStreamPrefix string) *cloudwatchPoller {
+	awsRegion string, config config) *cloudwatchPoller {
 	if metrics == nil {
 		metrics = newInputMetrics("", nil)
 	}
 
 	return &cloudwatchPoller{
-		numberOfWorkers:      numberOfWorkers,
-		apiSleep:             apiSleep,
-		region:               awsRegion,
-		logStreams:           logStreams,
-		logStreamPrefix:      logStreamPrefix,
-		startTime:            int64(0),
-		endTime:              int64(0),
-		workerSem:            awscommon.NewSem(numberOfWorkers),
 		log:                  log,
 		metrics:              metrics,
+		region:               awsRegion,
+		config:               config,
 		workersListingMap:    new(sync.Map),
 		workersProcessingMap: new(sync.Map),
+		// workRequestChan is unbuffered to guarantee that
+		// the worker and main loop agree whether a request
+		// was sent. workerResponseChan is buffered so the
+		// main loop doesn't have to block on the workers
+		// while distributing new data.
+		workRequestChan:  make(chan struct{}),
+		workResponseChan: make(chan workResponse, 10),
```

Comment:
Is this 10 chosen for a reason? Or just good practice?

Reply:
The choice to buffer it is because it lets the polling loop send multiple responses in one scheduler interval, but the choice of 10 is just an arbitrary small number. (Heuristically I think of it, buffer of 10 = ~90% less contention in the main loop than it would have had if it was synchronous. And I don't know exactly how much contention the synchronous case is, but surely eliminating 90% of it will avoid any bottleneck.)